mlir/lib/Dialect/PDL/IR/Builtins.cpp

```diff
@@ -39,6 +53,61 @@ mlir::Attribute addElemToArrayAttr(mlir::PatternRewriter &rewriter,
   values.push_back(element);
   return rewriter.getArrayAttr(values);
 }
+
+LogicalResult add(mlir::PatternRewriter &rewriter, mlir::PDLResultList &results,
+                  llvm::ArrayRef<mlir::PDLValue> args) {
+  assert(args.size() == 2 && "Expected 2 arguments");
+  auto lhsAttr = args[0].cast<Attribute>();
+  auto rhsAttr = args[1].cast<Attribute>();
+
+  // Integer
+  if (auto lhsIntAttr = dyn_cast_or_null<IntegerAttr>(lhsAttr)) {
+    auto rhsIntAttr = dyn_cast_or_null<IntegerAttr>(rhsAttr);
+    if (!rhsIntAttr || lhsIntAttr.getType() != rhsIntAttr.getType())
+      return failure();
+
+    auto integerType = lhsIntAttr.getType();
+
+    bool isOverflow;
+    llvm::APInt resultAPInt;
+    if (integerType.isUnsignedInteger()) {
+      resultAPInt =
+          lhsIntAttr.getValue().uadd_ov(rhsIntAttr.getValue(), isOverflow);
+    } else {
+      resultAPInt =
+          lhsIntAttr.getValue().sadd_ov(rhsIntAttr.getValue(), isOverflow);
+    }
+
+    if (isOverflow) {
+      return failure();
+    }
+
+    results.push_back(rewriter.getIntegerAttr(integerType, resultAPInt));
+    return success();
+  }
+
+  // Float
+  if (auto lhsFloatAttr = dyn_cast_or_null<FloatAttr>(lhsAttr)) {
+    auto rhsFloatAttr = dyn_cast_or_null<FloatAttr>(rhsAttr);
+    if (!rhsFloatAttr || lhsFloatAttr.getType() != rhsFloatAttr.getType())
```

Comment:
A nice improvement will be also result type reasoning but definitely for the future and not in this PR.
Like: f32 + f64 -> f64